README.md
# nu-react-native-button

[![npm package](https://img.shields.io/npm/v/@_nu/react-native-button.svg)](https://www.npmjs.org/package/@_nu/react-native-button)
[![github](https://img.shields.io/github/stars/nu-system/react-native-button.svg?style=social)](https://github.com/nu-system/react-native-button)

这是 NU 「 no-ui 」 组件库系统 nu-system，按钮组件 React native 实现。

![示意图](./assets/demo.png)

## 怎么用？

```bash
npm i @_nu/react-native-button
```

## 二次封装

选择 `NuButton` 意味着你选择的是一整套按钮解决方案。

因为不同的项目有不同的皮肤，在实际项目中使用需要二次封装。

这里默认采用的是 bootstrap 皮肤。

```JSX
import React from 'react';
import NuButton, {createNuButtonTheme} from '@_nu/react-native-button';

/**
 * 创建默认按钮主题
 */
createNuButtonTheme({
  levelColors: {
      default: '#343a40',
      secondary: '#6c757d',
      primary: '#007bff',
      warning: '#ffc107',
      danger: '#dc3545',
      success: '#22A745',
  }
});

/**
 * 创建 test 按钮主题
 */
createNuButtonTheme({
    name:'test',
    levelColors: {
        default: '#000000',
        primary: '#1976d2',
        secondary: 'rgb(220, 0, 78)',
        warning: '#ff9900',
        danger: 'red',
        success: 'green',
    }
});

export default NuButton;
```

## 使用

```JSX
import Button from "./components/Button";

const Page=()=>{    
    return (
     <div>
        <Button variant="ghost" size="large">一个大的幽灵按钮</Button>                
        <Button theme="test" variant="ghost" size="large">一个大的幽灵主按钮</Button>
     </div>     
    );
};
```

## Api

### NuButton

| props   | 类型 | 默认值 | 功能 |
|:-----|:-----:|:-----:|:-----:|
| theme |  'string' | 'default' | 按钮主题 |
| disabled |  boolean | - | 不可用按钮 |
| loading |  boolean | - | loading按钮 |
| capsule |  boolean | - | 圆角按钮|
| circle |  boolean | - | 正圆按钮 |
| level | 'default', 'primary', 'secondary', 'warning', 'danger', 'success' | `default` | 按钮等级 |
| variant | 'fill', 'ghost', 'link' | `fill` | 按钮变体 |
| size | \['large', 'default', 'middle', 'small' \] or `number` | `default` | 按钮大小 |
| before | 'element','function' | `null` | 放到文字之前 |
| after | 'element','function' | `null` | 放到文字之后 |

*注*: 虽然提供了 `size` 可以为 `number` 的接口，但是这个是非常不推荐使用的。你使用这种方式，意味着你定义的这个按钮样式是没法被人复用的。

### createNuButtonTheme 

| 参数   | 类型 | 默认值 | 功能 |
|:-----|:-----:|:-----:|:-----:|
| `name` |  `string` | `default` | 主题名称 |
| `defaultProps` |  `Object`| 见下面 | 默认属性 |
| `levelColors` |  `object` | 见下面 | 按钮主色列表 |
| `Wrap` |  `func node` | 见下面 | 按钮容器对象 |
| `WrapAndroid` |  `func node` | 见下面 | 安卓按钮容器对象 |
| `Content` |  `func node` | 见下面 | 内容器 |
| `Txt` |  `func node` | 见下面 | 文本容器 |
| `Loader` |  `func node` | 见下面 | 加载对象 |
| `wrapStyle` |  `object` | 见下面 | 主容器样式 |
| `contentStyle` |  `object` | 见下面 | 内容器样式 |
| `textStyle` |  `object` | 见下面 | 文本容器样式 |

### `defaultProps`

```JSX
defaultProps: {
    level: 'default',
    variant: 'fill'
}
```

这是默认会添加到按钮上的属性值。

### `Wrap, Content, Txt, Loader`

这四个对象拥有相同的API和使用方法，必须返回一个 `node` 对象。

```JSX
{
    WrapAndroid: function ({children, level, variant, ...otherProps}) {
        return <TouchableNativeFeedback
            background={TouchableNativeFeedback.SelectableBackground()} {...otherProps}>{children}</TouchableNativeFeedback>;
    },
    Wrap: function ({children, level, variant, ...otherProps}) {
        return <TouchableOpacity activeOpacity={0.5} {...otherProps}>{children}</TouchableOpacity>;
    },
    Txt: Text,
    Content: View,
    Loader: function ({level, variant, wrapStyle, contentStyle, textStyle}) {
        return (<ActivityIndicator animating={true} color={textStyle.color}/>);
    }
}
```

- level: 当下按钮的等级
- variant: 当下按钮的变体
- wrapStyle: 计算之后的主容器样式
- contentStyle: 计算之后的内容器样式
- textStyle:  计算之后的文本样式

#### 组件结构

```JSX
<Wrap>
    <Content>
        <Txt>Button</Txt>
    </Content>
</Wrap>
```

当 loading 的时候

```JSX
<Wrap>
    <Content>
        <Loader />
    </Content>
</Wrap>
```

### Api `wrapStyle` ,`contentStyle`, `textStyle`

这三个对象拥有相同的API和使用方法。

```JSX
{
    wrapStyle: {
        default: {
            position: 'relative',
            overflow: 'hidden',
            height: 40,
        },
        capsule: {
            borderRadius: 100
        },
        circle: {
            borderRadius: 100
        },
        large: {
            height: 48
        },
        middle: {
            height: 32
        },
        small: {
            height: 24
        },
        ghost: {
            borderWidth: 1
        },
        disabled: {
            opacity: 0.4
        }
    },
    textStyle: {
        default: {
            color: "#ffffff",
            fontSize: 16,
            textAlign: 'center',
            marginLeft: 8,
            marginRight: 8
        },
        large: {
            fontSize: 18
        },
        middle: {
            fontSize: 14
        },
        small: {
            fontSize: 12
        },
        customStyle: function ({warning, fill, style}) {
            if (warning && fill) {
                style.color = '#333333';
            }
            return style;
        }
    },
    contentStyle: {
        default: {
            width: '100%',
            height: '100%',
            alignItems: 'center',
            justifyContent: 'center',
            flexDirection: 'row'
        }
    }
}
```

除了 `default` 之外的所有 bool 属性，都会在实际渲染的时候和 `default` 样式合并然后经过 `customStyle` 再渲染。


#### customStyle

必须要返回 style 对象。

- level: 当下按钮的等级
- variant: 当下按钮的变体
- style: 当下计算好的样式


你甚至可以自定义任何 bool 属性。

```JSX
{
    wrapStyle:{
        h100:{
            height: 100
        }
    }
}
```

```JSX
<Button h100>高100的按钮</Button>
```

### `levelColors`

```JS
const levelColors= {
    default: '#343a40',
    primary: '#007bff',
    secondary: '#6c757d',
    warning: '#ffc107',
    danger: '#dc3545',
    success: '#22A745',
};
```

## FAQ

### 添加图标

```JSX
<Button before={<ActivityIndicator animating={true} color={'red'}/>}>图标在我左边</Button>
<Button after={<ActivityIndicator animating={true} color={'red'}/>}>图标在我右边</Button>
```

### 多行文本

```JSX
<Button>{({textStyle}) => {
    return (
        <View>
            <Text style={textStyle}>切换主题</Text>
            <Text style={textStyle}>「default</Text>
        </View>
    );
}}</Button>
```


### 渐变背景

```JSX
{    
    name:'test',
    Content: function ({children, style, level, variant}) {
        /* 修改 fill danger 按钮为渐变背景  */
        if (level === 'danger' && variant === 'fill') {
            return (
                <LinearGradient
                    colors={["red", "blue"]}
                    style={style}
                >{children}</LinearGradient>
            );
        } else {
            return (
                <View style={style}>{children}</View>
            );
        }
    },
}
```



CHANGELOG.md
## [0.1.6](https://github.com/nu-system/react-native-button/compare/v0.1.5...v0.1.6) (2019-08-22)



## [0.1.5](https://github.com/nu-system/react-native-button/compare/v0.1.4...v0.1.5) (2019-07-22)



## [0.1.4](https://github.com/nu-system/react-native-button/compare/v0.1.3...v0.1.4) (2019-07-18)



## [0.1.3](https://github.com/nu-system/react-native-button/compare/v0.1.2...v0.1.3) (2019-07-18)



## [0.1.2](https://github.com/nu-system/react-native-button/compare/v0.1.1...v0.1.2) (2019-07-18)


### Features

* 兼容按钮内容多子元素问题 ([351930e](https://github.com/nu-system/react-native-button/commit/351930e))



## [0.1.1](https://github.com/nu-system/react-native-button/compare/v0.1.0...v0.1.1) (2019-07-11)



# [0.1.0](https://github.com/nu-system/react-native-button/compare/v0.0.8...v0.1.0) (2019-07-05)



## [0.0.8](https://github.com/nu-system/react-native-button/compare/v0.0.7...v0.0.8) (2019-07-02)


### Bug Fixes

* 添加渐变背景 ([83227fb](https://github.com/nu-system/react-native-button/commit/83227fb))



## 0.0.7 (2019-07-02)





__tests__/setup.js
import "react-native-mock-render/mock";
import { JSDOM }  from "jsdom";
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });


const { document } = new JSDOM(``, {
  url: "https://example.com", // or whatever
}).window;
global.document = document;
global.window = document.defaultView;
Object.keys(document.defaultView).forEach(property => {
  if (typeof global[property] === "undefined") {
    global[property] = document.defaultView[property];
  }
});
function suppressDomErrors() {
  const suppressedErrors = /(React does not recognize the.*prop on a DOM element|Unknown event handler property|is using uppercase HTML|Received `true` for a non-boolean attribute `accessible`|The tag.*is unrecognized in this browser|PascalCase)/;
  // eslint-disable-next-line no-console
  const realConsoleError = console.error;
  // eslint-disable-next-line no-console
  console.error = message => {
    if (message.match(suppressedErrors)) {
      return;
    }
    realConsoleError(message);
  };
}
suppressDomErrors()

__tests__/basic/size.test.js
import 'react-native';
import React from 'react';
import NuButton from '../../lib/index';
import {shallow} from 'enzyme';

const wrapSize={
    default:40,
    large: 48,
    middle: 32,
    small: 24
}
const textSize={
    default:16,
    large: 18,
    middle: 14,
    small: 12
}
describe("枚举size属性，判断结果是否正确",()=>{
    // 'large', 'default', 'middle', 'small'
    it(`default 状态`,()=>{
        const tmp= shallow(<NuButton>default</NuButton>)
        const {theme , style, Pre, Append, ...userProps}=tmp.props()
        expect(userProps.wrapStyle.height).toBe(wrapSize.default)
        expect(userProps.textStyle.fontSize).toBe(textSize.default)
    })
    it(`large 状态`,()=>{
        const tmp= shallow(<NuButton size="large">large</NuButton>)
        const {theme , style, Pre, Append, ...userProps}=tmp.props()
        expect(userProps.wrapStyle.height).toBe(wrapSize.large)
        expect(userProps.textStyle.fontSize).toBe(textSize.large)
    })
    it(`middle 状态`,()=>{
        const tmp= shallow(<NuButton size="middle">middle</NuButton>)
        const {theme , style, Pre, Append, ...userProps}=tmp.props()
        expect(userProps.wrapStyle.height).toBe(wrapSize.middle)
        expect(userProps.textStyle.fontSize).toBe(textSize.middle)
    })
    it(`small 状态`,()=>{
        const tmp= shallow(<NuButton size="small">small</NuButton>)
        const {theme , style, Pre, Append, ...userProps}=tmp.props()
        expect(userProps.wrapStyle.height).toBe(wrapSize.small)
        expect(userProps.textStyle.fontSize).toBe(textSize.small)
    })
})
describe("自定义size是可以的，必须是数字",()=>{
    it("custom size is number",()=>{
        const size=40
        const tmp=shallow(<NuButton size={size}>{size}</NuButton>)
        expect(tmp.prop("wrapStyle").height).toBe(size)
    })
    it("custom size is not number",()=>{
        const size="custom"
        const oriError=console.error
        const warningError=new Error("warning")
        console.error=()=>{
            throw warningError
        }
        expect(()=>{<NuButton size={size}>{size}</NuButton>}).toThrowError(warningError)
        console.error=oriError
    })
})


__tests__/render/ios.test.js
import 'react-native';
import React from 'react';
import NuButton from '../../lib/index';
import {shallow,mount} from 'enzyme';

describe("ios 容器测试",()=>{
    it("ios render correctly",()=>{
        const fb= shallow(<NuButton>default</NuButton>)
        expect(fb.find("Wrap").find("Text").prop("children")).toBe("default")
    })
    it("ios touchopacity",()=>{
        const fb= mount(<NuButton>default</NuButton>)
        expect(fb.find("TouchableOpacity").debug()).not.toBe("")
    })
})

demo/App.js
import React, { useState } from 'react';
import {
  StyleSheet, Text, View, TouchableHighlight, Image, ActivityIndicator,
} from 'react-native';
import { Row, Col, Wrap } from './components/Grid/index.js';
import Button from './components/Button/index.js';

export default function App() {
  const [theme, setTheme] = useState('default');
  return (
    <View style={{
      height: '100%',
    }}
    >
      <Text>Hello world</Text>
      <View style={{ height: 50 }} />
      {/* <Wrap> */}
      {/*    <Text style={styles.col}>Open up App.js to start</Text> */}
      {/* </Wrap> */}
      {/* <Row> */}
      {/*    <Col span={1}><Text style={styles.col}>1</Text></Col> */}
      {/*    <Col span={1}><Text style={styles.col}>2</Text></Col> */}
      {/*    <Col span={1}><Text style={styles.col}>3</Text></Col> */}
      {/*    <Col span={1}><Text style={styles.col}>4</Text></Col> */}
      {/*    <Col span={1}><Text style={styles.col}>5</Text></Col> */}
      {/*    <Col span={1}><Text style={styles.col}>6</Text></Col> */}
      {/* </Row> */}
      {/* <Row> */}
      {/*    <Col span={2}><Text style={styles.col}>Hello1</Text></Col> */}
      {/*    <Col span={2} offset={2}><Text style={styles.col}>Hello2</Text></Col> */}
      {/*    <Col span={4} widthToInt><Text style={styles.col}>Hello1</Text></Col> */}
      {/*    <Col span={2}><Text style={styles.col}>Hello1</Text></Col> */}
      {/*    <Col span={4}><Text style={styles.col}>Hello1</Text></Col> */}
      {/*    <Col span={2}><Text style={styles.col}>Hello1</Text></Col> */}
      {/* </Row> */}
      {/* <Row> */}
      {/*    <Col span={3}><Text style={styles.col}>Hello1</Text></Col> */}
      {/*    <Col span={3}><Text style={styles.col}>Hello2</Text></Col> */}
      {/* </Row> */}
      <Row style={{
        marginBottom: 4,
      }}
      >
        <Col span={2}>
          <Button theme={theme} before={<ActivityIndicator animating color="red" />}>fill</Button>
        </Col>
        <Col span={2}>
          <Button theme={theme} variant="ghost">Ghost</Button>
        </Col>
        <Col span={2}>
          <Button theme={theme} variant="link">Link</Button>
        </Col>
      </Row>
      <Row style={{
        marginBottom: 20,
      }}
      >
        <Col span={2}>
          <Button style={{ marginBottom: 4 }} theme={theme} level="primary">primary</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} level="secondary">secondary</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} level="warning">warning</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} level="danger">danger</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} level="success">success</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} capsule>capsule</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} loading>default loading</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} disabled>disabled</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} size="large">large</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} size="middle">middle</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} size="small">small</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} circle>+</Button>
        </Col>
        <Col span={2}>

          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" level="primary">primary</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" level="secondary">secondary</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" level="warning">warning</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" level="danger">danger</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" level="success">success</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" capsule>capsule2</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" loading>loading</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" disabled>disabled</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" size="large">large</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" size="middle">middle</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" size="small">small</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="ghost" circle>+</Button>
        </Col>
        <Col span={2}>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" level="primary">primary</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" level="secondary">secondary</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" level="warning">warning</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" level="danger">danger</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" level="success">success</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" capsule>capsule</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" loading>loading</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" disabled>disabled</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" size="large">large</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" size="middle">middle</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link" size="small">small</Button>
          <Button style={{ marginBottom: 4 }} theme={theme} variant="link">+</Button>
        </Col>
      </Row>
      <Wrap>
        <Button
          style={{ marginBottom: 4 }}
          theme={theme}
          onPress={() => {
            setTheme(theme === 'default' ? 'test' : 'default');
          }}
        >
          {({ textStyle }) => (
            <View>
              <Text style={textStyle}>切换主题</Text>
              <Text style={textStyle}>{theme}</Text>
            </View>
          )}
        </Button>
      </Wrap>
      <View style={{
        alignItems: 'flex-start',
      }}
      >
        <View style={{
          backgroundColor: 'red',
          height: 50,
        }}
        >
          <Text style={{
            paddingLeft: 10,
            paddingRight: 10,
          }}
          >
asdf
          </Text>
        </View>
      </View>
      <View style={{
        alignItems: 'flex-start',
      }}
      >
        <Button>123</Button>
      </View>

    </View>
  );
}


jest.config.js
// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  // All imported modules in your tests should be mocked automatically
  // automock: false,

  // Stop running tests after `n` failures
  // bail: 0,

  // Respect "browser" field in package.json when resolving modules
  // browser: false,

  // The directory where Jest should store its cached dependency information
  // cacheDirectory: "/private/var/folders/f3/p_8cqzpx6q35q4wkwbzdnf1r0000gn/T/jest_dx",

  // Automatically clear mock calls and instances between every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  // collectCoverage: false,

  // An array of glob patterns indicating a set of files for which coverage information should be collected
  // collectCoverageFrom: null,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // An array of regexp pattern strings used to skip coverage collection
  coveragePathIgnorePatterns: [
    "/node_modules/",
    "/demo/"
  ],

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  // An object that configures minimum threshold enforcement for coverage results
  // coverageThreshold: null,

  // A path to a custom dependency extractor
  // dependencyExtractor: null,

  // Make calling deprecated APIs throw helpful error messages
  // errorOnDeprecated: false,

  // Force coverage collection from ignored files using an array of glob patterns
  // forceCoverageMatch: [],

  // A path to a module which exports an async function that is triggered once before all test suites
  // globalSetup: null,

  // A path to a module which exports an async function that is triggered once after all test suites
  // globalTeardown: null,

  // A set of global variables that need to be available in all test environments
  // globals: {},

  // An array of directory names to be searched recursively up from the requiring module's location
  // moduleDirectories: [
  //   "node_modules"
  // ],

  // An array of file extensions your modules use
  // moduleFileExtensions: [
  //   "js",
  //   "json",
  //   "jsx",
  //   "ts",
  //   "tsx",
  //   "node"
  // ],

  // A map from regular expressions to module names that allow to stub out resources with a single module
  // moduleNameMapper: {},

  // An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
  modulePathIgnorePatterns: [
    "/demo/"
  ],

  // Activates notifications for test results
  // notify: false,

  // An enum that specifies notification mode. Requires { notify: true }
  // notifyMode: "failure-change",

  // A preset that is used as a base for Jest's configuration
  preset: 'react-native',

  // Run tests from one or more projects
  // projects: null,

  // Use this configuration option to add custom reporters to Jest
  reporters: ["default", "<rootDir>__tests__/my-custom-reporter.js"],

  // Automatically reset mock state between every test
  // resetMocks: false,

  // Reset the module registry before running each individual test
  // resetModules: false,

  // A path to a custom resolver
  // resolver: null,

  // Automatically restore mock state between every test
  // restoreMocks: false,

  // The root directory that Jest should scan for tests and modules within
  // rootDir: null,

  // A list of paths to directories that Jest should use to search for files in
  // roots: [
  //   "<rootDir>"
  // ],

  // Allows you to use a custom runner instead of Jest's default test runner
  // runner: "jest-runner",

  // The paths to modules that run some code to configure or set up the testing environment before each test
  // setupFiles: [],

  // A list of paths to modules that run some code to configure or set up the testing framework before each test
  setupFilesAfterEnv: ["<rootDir>__tests__/setup.js"],

  // A list of paths to snapshot serializer modules Jest should use for snapshot testing
  // snapshotSerializers: [],

  // The test environment that will be used for testing
  testEnvironment: "jsdom",

  // Options that will be passed to the testEnvironment
  // testEnvironmentOptions: {},

  // Adds a location field to test results
  // testLocationInResults: false,

  // The glob patterns Jest uses to detect test files
  testMatch: [
    "**/__tests__/**/*.test.[jt]s?(x)",
    "**/?(*.)+(spec|test).[tj]s?(x)"
  ],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  testPathIgnorePatterns: [
    "/node_modules/",
    "demo"
  ],

  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: null,

  // This option allows use of a custom test runner
  // testRunner: "jasmine2",

  // This option sets the URL for the jsdom environment. It is reflected in properties such as location.href
  // testURL: "http://localhost",

  // Setting this value to "fake" allows the use of fake timers for functions such as "setTimeout"
  // timers: "real",

  // A map from regular expressions to paths to transformers
  // transform: null,

  // An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
  transformIgnorePatterns: [
    "node_modules/(?!(jest-)?react-native)"
  ],

  // An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
  // unmockedModulePathPatterns: undefined,

  // Indicates whether each individual test should be reported during the run
  // verbose: null,

  // An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
  // watchPathIgnorePatterns: [],

  // Whether to use watchman for file crawling
  // watchman: true,
};


lib/index.js
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import {
  Text,
  View,
  ActivityIndicator,
  TouchableNativeFeedback,
  TouchableOpacity,
  Platform,
  StyleSheet,
} from 'react-native';
import objectAssignDeep from './objectAssignDeep';

/**
 *
 * [私有接口]
 * @private
 */
// eslint-disable-next-line
const _apiPrivate = {
  isAndroid: Platform.OS === 'android',
  themes: {},
  getButtonInfoByProps({ styleConfig, useProps }) {
    const {
      levelColors, wrapStyle, textStyle, contentStyle,
    } = styleConfig;
    const styleProps = { ...useProps };
    const levelColor = levelColors[styleProps.level];
    const domsStyle = {
      wrap: { ...wrapStyle.default },
      text: { ...textStyle.default },
      content: { ...contentStyle.default },
    };

    /**
     * 等级变成 bool 类型
     */
    if (styleProps.level !== 'default') {
      styleProps[styleProps.level] = true;
    }

    /**
     * 将 size 变成 bool props
     */
    if (styleProps.size) {
      styleProps[styleProps.size] = true;
    }

    /**
     * 将变体和 level 变成 bool props
     */
    styleProps[styleProps.variant] = true;


    /**
     * circle 按钮 高宽一致
     */
    if (styleProps.circle) {
      domsStyle.wrap.width = domsStyle.wrap.height;
    }

    /**
     * 用户可以给 size 传数字
     */
    if (typeof styleProps.size === 'number') {
      domsStyle.wrap.height = styleProps.size;
    }

    if (styleProps.variant === 'fill') {
      // fill 按钮背景颜色等于主色
      domsStyle.wrap.backgroundColor = levelColor;
    } else if (styleProps.variant === 'ghost') {
      // ghost 按钮边框和文字颜色等于主色
      domsStyle.wrap.borderColor = levelColor;
      domsStyle.text.color = levelColor;
    } else if (styleProps.variant === 'link') {
      // link 按钮文字颜色等于主色
      domsStyle.text.color = levelColor;
    }

    /**
     * 将 prop 类型的样式合并到基础到样式中
     */
    const domsConfig = {
      wrap: wrapStyle,
      text: textStyle,
      content: contentStyle,
    };
    // eslint-disable-next-line no-restricted-syntax
    for (const [propKey, propValue] of Object.entries(styleProps)) {
      // eslint-disable-next-line no-restricted-syntax
      for (const [domKey, domConfig] of Object.entries(domsConfig)) {
        if (propValue === true && domConfig[propKey]) {
          domsStyle[domKey] = { ...domsStyle[domKey], ...domConfig[propKey] };
        }
      }
    }

    /**
     * 将 用户自定义的样式样式合并到基础到样式中
     */
    const customStyles = {
      wrap: wrapStyle.customStyle,
      text: textStyle.customStyle,
      content: contentStyle.customStyle,
    };
    // eslint-disable-next-line no-restricted-syntax
    for (const [key, customFunc] of Object.entries(customStyles)) {
      if (typeof customFunc === 'function') {
        const defaultStyle = domsStyle[key];
        domsStyle[key] = customFunc({
          levelColor,
          style: { ...defaultStyle },
          ...styleProps,
        }) || defaultStyle;
      }
    }

    return {
      levelColor,
      wrapStyle: { ...domsStyle.wrap },
      textStyle: { ...domsStyle.text },
      contentStyle: { ...domsStyle.content },
    };
  },

  renderChildren({
    buttonInfo, Txt, before, after, children,
  }) {
    /**
     * 如果子元素是个function 那么直接执行
     */
    if (typeof children === 'function') {
      return children({ ...buttonInfo });
    }

    let allChildrenIsString = true;
    React.Children.map(children, (element) => {
      if (typeof element !== 'string') {
        allChildrenIsString = false;
      }
    });

    /**
     * 如果所有的子元素都是字符串的时候放到 text 组件中
     */
    if (allChildrenIsString) {
      return (
      // eslint-disable-next-line react/jsx-filename-extension
        <Fragment>
          {typeof before === 'function' ? before({ ...buttonInfo }) : before}
          <Txt {...buttonInfo} style={buttonInfo.textStyle}>{children}</Txt>
          {typeof after === 'function' ? after({ ...buttonInfo }) : after}
        </Fragment>
      );
    }
    /**
     * 子元素除了 function 和 全部都是字符串之外的逻辑就直接输出 children
     */
    return children;
  },
};

/**
 * [默认配置]
 * @private
 */
// eslint-disable-next-line
const _defaultConfig = {
  name: 'default',
  defaultProps: {
    level: 'default',
    variant: 'fill',
  },
  levelColors: {
    default: '#343a40',
    primary: '#007bff',
    secondary: '#6c757d',
    warning: '#ffc107',
    danger: '#dc3545',
    success: '#22A745',
  },
  wrapStyle: {
    default: {
      position: 'relative',
      overflow: 'hidden',
      height: 40,
    },
    capsule: {
      borderRadius: 100,
    },
    circle: {
      borderRadius: 100,
    },
    large: {
      height: 48,
    },
    middle: {
      height: 32,
    },
    small: {
      height: 24,
    },
    ghost: {
      borderWidth: 1,
    },
    disabled: {
      opacity: 0.4,
    },
  },
  textStyle: {
    default: {
      color: '#ffffff',
      fontSize: 16,
      textAlign: 'center',
      marginLeft: 8,
      marginRight: 8,
    },
    large: {
      fontSize: 18,
    },
    middle: {
      fontSize: 14,
    },
    small: {
      fontSize: 12,
    },
  },
  contentStyle: {
    default: {
      height: '100%',
      alignItems: 'center',
      justifyContent: 'center',
      flexDirection: 'row',
    },
  },
  Content: View,
  // eslint-disable-next-line
  WrapAndroid({children, style, ...otherProps}) {
    return (
      <View style={style}>
        <TouchableNativeFeedback
          background={TouchableNativeFeedback.SelectableBackground()}
          {...otherProps}
        >
          {children}
        </TouchableNativeFeedback>
      </View>
    );
  },
  // eslint-disable-next-line
  Wrap({children, ...otherProps}) {
    return <TouchableOpacity activeOpacity={0.5} {...otherProps}>{children}</TouchableOpacity>;
  },
  // eslint-disable-next-line
  Loader({textStyle}) {
    // eslint-disable-next-line
    return (<ActivityIndicator animating color={textStyle.color}/>);
  },
  Txt: Text,
};

const NuButton = ({
  theme,
  // eslint-disable-next-line react/prop-types
  style,
  ...userProps
}) => {
  // eslint-disable-next-line
  const _config = _apiPrivate.themes[theme] ? _apiPrivate.themes[theme] : _apiPrivate.themes.default;
  const {
    Wrap, wrapProps, Loader, Content, Txt, defaultProps, ...styleConfig
  } = _config;
  const useProps = { ...defaultProps, ...userProps};
  const buttonInfo = _apiPrivate.getButtonInfoByProps({ styleConfig, useProps });
  const {
    loading, disabled, children, before, after, ...leftProps
  } = useProps;
  return (
    <Wrap
      {...buttonInfo}
      style={StyleSheet.flatten([buttonInfo.wrapStyle, style])}
      disabled={disabled || loading}
      {...wrapProps}
      {...leftProps}
    >
      <Content {...buttonInfo} style={buttonInfo.contentStyle} {...leftProps}>
        {loading ? Loader({ ...buttonInfo }) : _apiPrivate.renderChildren({
          Txt,
          children,
          before,
          after,
          buttonInfo,
        })}
      </Content>
    </Wrap>
  );
};

const ButtonPropTypes = {
  theme: PropTypes.string,
  /**
   * 不可用状态
   */
  // eslint-disable-next-line
  disabled: PropTypes.bool,
  /**
   * loading 状态
   */
  // eslint-disable-next-line
  loading: PropTypes.bool,
  /**
   *  按钮等级
   */
  // eslint-disable-next-line
  level: PropTypes.oneOf(['default', 'primary', 'secondary', 'warning', 'danger', 'success']),
  /**
   *  按钮大小
   */
  // eslint-disable-next-line
  size: PropTypes.oneOfType([
    PropTypes.oneOf(['large', 'middle', 'small']),
    PropTypes.number,
  ]),
  /**
   *  按钮变体
   */
  // eslint-disable-next-line
  variant: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.oneOf(['fill', 'ghost', 'link']),
  ]),
  /**
   * 胶囊按钮，左右都是圆角的按钮
   */
  // eslint-disable-next-line
  capsule: PropTypes.bool,
  /**
   * 纯圆按钮
   */
  // eslint-disable-next-line
  circle: PropTypes.bool,
  /**
   * 放到文字之前的元素
   */
  before: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
  ]),
  /**
   * 放到文字之后的元素
   */
  after: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
  ]),
};

NuButton.propTypes = ButtonPropTypes;
NuButton.defaultProps = {
  theme: 'default',
  before: null,
  after: null,
};

/**
 * 创建主题
 * @param {string} name - 主题名称
 * @param {object} levelColors - 按钮主色
 * @param {node} Wrap - 按钮容器对象
 * @param {node} WrapAndroid - 安卓按钮容器对象
 * @param {object} wrapStyle - 主容器样式
 * @param {object} Content - 内容器
 * @param {object} contentStyle - 内容器样式
 * @param {object} Txt - 文本容器
 * @param {object} textStyle - 文本容器样式
 * @param {object} Loader - 加载对象
 */
function createNuButtonTheme({ name = 'default', ...otherConfig }) {
  const newConfig = objectAssignDeep({}, _defaultConfig, otherConfig);
  if (_apiPrivate.isAndroid) {
    newConfig.Wrap = newConfig.WrapAndroid;
  }
  _apiPrivate.themes[name] = newConfig;
}
createNuButtonTheme(_defaultConfig);
export default NuButton;
export { createNuButtonTheme, ButtonPropTypes };
